week5/Q1.py

```diff
@@ -0,0 +1,149 @@
+class OrientedGraph:
+	'''
+	Implementation of oriented graph. All of the nodes have unique values.
+
+	Fields:
+		nodes: set of Nodes, all nodes of the graph
+	'''
+	def __init__(self):
+		self.nodes = set()
+
+	def add(self, value, parent_value = None):
+		'''
+		Adds new Node to the OrientedGraph or updates existing ones.
+
+		Args:
+			value: str
+			parent_value: str
+		'''
+		node = None
+		parent = None
+		for inode in self.nodes:
+			if inode.value == value:
+				node = inode
+			if inode.value == parent_value:
+				parent = inode
+
+		if parent_value:
+			if node:
+				node.parents = node.parents.union({parent})
+				parent.children = parent.children.union({node})
+			else:
+				node = Node(value, parents = {parent})
+				parent.children = parent.children.union({node})
+				self.nodes.add(node)
+		else:
+			if not node:
+				node = Node(value)
+				self.nodes.add(node)
+
+
+class Node:
+	'''
+	Element of OrientedGraph.
+
+	Fields:
+		value: str
+		parents: set of Nodes, from which Node has incoming edges
+		children: set of Nodes, to which Node has outcoming edges
+	'''
+	def __init__(self, value = None, parents = set(), children = set()):
+		self.value = value
+		self.parents = parents
+		self.children = children
+
+
+def _drop_empty_words(words):
+	'''
+	Removes empty elements from list.
+
+	Args:
+		words: list
+
+	Returns:
+		list
+	'''
+	words = [word for word in words if word != '']
+	return words
+
+
+def _get_ordered_letters(dictionary, letter_graph):
+	'''
+	Builds oriented graph with no cycles which represents hierarchy of letters.
+
+	Args:
+		dictionary: list, words in lexicogrpahic order
+		letter_graph: OrientedGraph
+
+	Returns:
+		OrientedGraph
+	'''
+	same_prefix_words = []
+	last_letter = None
+	for word in dictionary:
+		if last_letter == word[0]:
+			same_prefix_words.append(word[1:])
+		else:
+			same_prefix_words = _drop_empty_words(same_prefix_words)
+			if same_prefix_words != []:
+				letter_graph = _get_ordered_letters(same_prefix_words, letter_graph)
+			same_prefix_words = [word[1:]]
+			letter_graph.add(word[0], last_letter)
+			last_letter = word[0]
+
+	same_prefix_words = _drop_empty_words(same_prefix_words)
+	if same_prefix_words != []:
+		letter_graph = _get_ordered_letters(same_prefix_words, letter_graph)
+	return letter_graph
+
+
+def _topological_sort(graph):
+	'''
+	Builds a possible alphabet from given letter hierarchy.
+
+	Args:
+		graph: OrientedGraph
+
+	Returns:
+		list
+	'''
+	sorted_elements = []
+	for node in graph.nodes:
+		if node.parents == set():
+			sorted_elements.append(node.value)
+	graph.nodes = {node for node in graph.nodes if node.value not in sorted_elements}
+	for node in graph.nodes:
+		node.parents = {parent for parent in node.parents if parent.value not in sorted_elements}
+	if graph.nodes:
+		sorted_elements += _topological_sort(graph)
```

Comment:
Optimization:
Here you change the graph and work recursively.
This certainly works, but there is some cost: for example, changing the data structure each time.

For a more performant solution, you can use something that:
- is iterative
- does not modify the graph

Two suggestions: instead of modifying the graph, keep track of which nodes you have already *remove* from it and instead of looking for nodes without parents, look for nodes without parents that have no already been removed. Intuitively, if the parents of `L` are `A` and 'X', once both have been taken out, you can safely take out `L` as well. As you do that, you will notice that you no longer need to do this recursively, because the graph has not changed, and there is no different between removing those that have no parents and those that have only parents that have already been processed. Besides avoiding the recursion, this solution may also finish in fewer steps: as you iterate over the remaining nodes, in a single iteration if you remove first `A` and then `X` from the example above, you can directly remove `L` without having to iterate again (assuming that you actually considered the nodes in that order; in the worst case, however, it is the same).